libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -199,19 +199,29 @@ void GCS_MAVLINK::send_battery_status(const uint8_t instance) const
     float temp;
     bool got_temperature = battery.get_temperature(temp, instance);
 
-    // ensure we always send a voltage estimate to the GCS, because not all battery monitors monitor individual cells
-    // as a work around for this we create a set of fake cells to be used if the backend doesn't provide direct monitoring
-    // the GCS can then recover the pack voltage by summing all non ignored cell values. Because this is looped we can
-    // report a pack up to 655.34 V with this method
-    AP_BattMonitor::cells fake_cells;
-    if (!battery.has_cell_voltages(instance)) {
+    // prepare array of individual cell voltage
+    uint16_t cell_volts[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN];
+    if (battery.has_cell_voltages(instance)) {
+        const AP_BattMonitor::cells& batt_cells = battery.get_cell_voltages(instance);
+        for (uint8_t i = 0; i < ARRAY_SIZE(batt_cells.cells); i++) {
+            if (i < MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN) {
+                cell_volts[i] = batt_cells.cells[i];
+            } else {
+                // 10th cell reports the lowest voltage of the last 3 cells
+                cell_volts[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN-1] = MIN(cell_volts[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN-1], batt_cells.cells[i]);
```

Comment:
`cell_volts` is never initialized that I can see, which is fine and reasonable when data is just being copied, however that means when you get to your 10th cell `cell_volts[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN-1]` is uninitialized and is random stack data. It may be working now but it's not guaranteed, at least not any way that I can see? You can just set the last cell to be 65535 before the loop starts....

Reply:
No, I don't think so.  Perhaps I'm missing something but I think within the for loop, cell_volts[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN-1] is copied from batt_cells.cells[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN-1].

Then on the next two iterations we may update this same cell again with lower voltages.

Reply:
Ah, yup, your correct. It does sorta point at this being a confusing structure though. I've left a suggestion below which swaps to memcpy, and a shorter for loop that I think is clearer. It compiles locally, but I don't have any hardware that I can currently connect to actually validate the behavior on.

Reply:
Great, I'll commit that and give it a test.